### DISCUSSION TITLE
Block uses a static buffer.

### DISCUSSION THREAD

**Discussion by Jdance-Media** [SmartlyDressedGames/Unturned-3.x-Community#4170] [Ideas]
Been using Block for a long while now for storing data instead of using JSON. Unfortunately, I'm currently unable to use it on another thread due to the static buffer. Making Block threadsafe would simply improve performance for plugins that use it, especially if the plugin involves something like a virtual garage. Perhaps an okay solution to this would be to have the option of passing in your own buffer.

**Comment by GazziFX** [comment]
Not sure why you wanna to use it in another thread, Task will continue in current synchronization context so you don't care about different thread. If it really needs to be in separate thread you can mark static Block fields with [ThreadStatic]

**Comment by sunnamed434** [comment]
You mean that you need to return to the main thread?

**Comment by Jdance-Media** [comment]
> Not sure why you wanna to use it in another thread, Task will continue in current synchronization context so you don't care about different thread. If it really needs to be in separate thread you can mark static Block fields with [ThreadStatic]

I wish to use another thread because reading and writing data to files makes sense to have on another thread rather than the main thread. Having a static buffer also becomes an annoyance when I want to write several Block files (.DAT) on multiple threads at once. I have tried to read/write using Block several times before on a different thread and I have been met consistently met with errors regarding it.

**Reply by GazziFX** [reply]
You can serialize blocks on main thread and copy to another buffer to write to disk asynchronously

**Comment by SDGNelson** [comment]
Someday I'd like to consolidate the "Block"/"River" classes into something more robust. Marking it ThreadStatic as GazziFX mentioned might be the best option for now.

**Reply by andriivitiv** [reply]
Btw I would suggest replacing `BitConverter.GetBytes` with `BitConverter.TryWriteBytes` to avoid unnecessary array allocation. E.g. [this line](https://github.com/Unturned-Datamining/Unturned-Datamining/blob/6f1255fb44a2449b3a7cd0e0305a67d76bef1819/Assembly-CSharp/SDG.Unturned/LevelObjects.cs#L707) alone (called 4096 times) will make GC work hard.

Suggested code:

```cs
class Block
{
    public void writeSingle(float value)
    {
        const int size = sizeof(float);
        var span = buffer.AsSpan(step, size); // will throw on overflow
        BitConverter.TryWriteBytes(span, value);
        step += size;
    }

    // ...
}

class River
{
    public void writeSingle(float value)
    {
        if (block != null)
        {
            block.writeSingle(value);
            return;
        }

        const int size = sizeof(float);
        Span<byte> span = stackalloc byte[size];
        BitConverter.TryWriteBytes(span, value);
        stream.Write(span);
        water += size;
    }

    // ...
}
```

**Reply by SDGNelson** [reply]
Thanks @CyberAndrii, that is a nice improvement! I don't think that was available previously. It will be nice to have some of the newer non-allocating methods available.
